File: backend/app/scrapers/acled.py
import logging
from datetime import datetime, timedelta

import httpx

from app.core.config import settings
from app.schemas.event import EventCreate
from app.scrapers.base import BaseScraper

logger = logging.getLogger(__name__)

ACLED_BASE_URL = "https://api.acleddata.com/acled/read"

EVENT_TYPE_MAP = {
    "Battles": "battle",
    "Explosions/Remote violence": "explosion",
    "Violence against civilians": "violence_against_civilians",
    "Protests": "protest",
    "Riots": "riot",
    "Strategic developments": "strategic_development",
}


class AcledScraper(BaseScraper):
    source_name = "acled"

    def __init__(self, days_back: int = 7):
        self.days_back = days_back

    async def fetch_raw(self) -> list[dict]:
        date_from = (datetime.utcnow() - timedelta(days=self.days_back)).strftime(
            "%Y-%m-%d"
        )
        date_to = datetime.utcnow().strftime("%Y-%m-%d")

        all_data = []
        page = 1

        async with httpx.AsyncClient(timeout=60) as client:
            while True:
                params = {
                    "key": settings.ACLED_API_KEY,
                    "email": settings.ACLED_EMAIL,
                    "event_date": f"{date_from}|{date_to}",
                    "event_date_where": "BETWEEN",
                    "limit": 500,
                    "page": page,
                }

                for attempt in range(3):
                    try:
                        resp = await client.get(ACLED_BASE_URL, params=params)
                        resp.raise_for_status()
                        break
                    except httpx.HTTPError as e:
                        logger.warning(
                            f"[acled] HTTP error (attempt {attempt + 1}): {e}"
                        )
                        if attempt == 2:
                            raise

                body = resp.json()
                data = body.get("data", [])
                if not data:
                    break

                all_data.extend(data)
                page += 1

                if len(data) < 500:
                    break

        return all_data

    def normalize(self, raw: dict) -> EventCreate | None:
        lat = raw.get("latitude")
        lon = raw.get("longitude")
        if not lat or not lon:
            return None

        try:
            lat_f = float(lat)
            lon_f = float(lon)
        except (ValueError, TypeError):
            return None

        if lat_f == 0 and lon_f == 0:
            return None

        raw_type = raw.get("event_type", "")
        event_type = EVENT_TYPE_MAP.get(raw_type, raw_type.lower().replace(" ", "_"))

        actors = []
        if raw.get("actor1"):
            actors.append(raw["actor1"])
        if raw.get("actor2"):
            actors.append(raw["actor2"])

        fatalities = 0
        try:
            fatalities = int(raw.get("fatalities", 0))
        except (ValueError, TypeError):
            pass

        event_date_str = raw.get("event_date", "")
        try:
            event_date = datetime.strptime(event_date_str, "%Y-%m-%d")
        except ValueError:
            return None

        location_parts = []
        if raw.get("location"):
            location_parts.append(raw["location"])
        if raw.get("admin1"):
            location_parts.append(raw["admin1"])

        return EventCreate(
            source="acled",
            source_id=str(raw.get("data_id", "")),
            event_type=event_type,
            sub_event_type=raw.get("sub_event_type"),
            title=f"{raw.get('sub_event_type', raw_type)} in {raw.get('location', 'Unknown')}",
            description=raw.get("notes"),
            url=raw.get("source_url"),
            latitude=lat_f,
            longitude=lon_f,
            location_name=", ".join(location_parts) if location_parts else None,
            country=raw.get("country"),
            region=raw.get("region"),
            event_date=event_date,
            fatalities=fatalities,
            actors=actors if actors else None,
            tags=[raw_type] if raw_type else None,
        )

File: backend/src/scrapers/acled.ts
import { config } from "../config";
import { EventCreate } from "../models/types";
import { bulkUpsert } from "../services/eventService";
import { broadcast } from "../services/broadcast";

const ACLED_BASE_URL = "https://api.acleddata.com/acled/read";

const EVENT_TYPE_MAP: Record<string, string> = {
  Battles: "battle",
  "Explosions/Remote violence": "explosion",
  "Violence against civilians": "violence_against_civilians",
  Protests: "protest",
  Riots: "riot",
  "Strategic developments": "strategic_development",
};

export async function scrapeAcled(daysBack = 7): Promise<number> {
  if (!config.acledApiKey || !config.acledEmail) {
    console.log("[acled] No API key configured, skipping");
    return 0;
  }

  console.log("[acled] Starting scrape...");
  const now = new Date();
  const from = new Date(now.getTime() - daysBack * 86400000);
  const dateFrom = from.toISOString().slice(0, 10);
  const dateTo = now.toISOString().slice(0, 10);

  const allData: any[] = [];
  let page = 1;

  while (true) {
    const url = new URL(ACLED_BASE_URL);
    url.searchParams.set("key", config.acledApiKey);
    url.searchParams.set("email", config.acledEmail);
    url.searchParams.set("event_date", `${dateFrom}|${dateTo}`);
    url.searchParams.set("event_date_where", "BETWEEN");
    url.searchParams.set("limit", "500");
    url.searchParams.set("page", String(page));

    let resp: Response;
    for (let attempt = 0; attempt < 3; attempt++) {
      try {
        resp = await fetch(url.toString());
        if (resp.ok) break;
      } catch (e) {
        console.warn(`[acled] Attempt ${attempt + 1} failed:`, e);
        if (attempt === 2) throw e;
        await new Promise((r) => setTimeout(r, 2000 * (attempt + 1)));
      }
    }

    const body: any = await resp!.json();
    const data = body.data || [];
    if (data.length === 0) break;

    allData.push(...data);
    page++;
    if (data.length < 500) break;
  }

  console.log(`[acled] Fetched ${allData.length} raw items`);

  const events: EventCreate[] = [];
  for (const raw of allData) {
    const ev = normalizeAcled(raw);
    if (ev) events.push(ev);
  }

  console.log(`[acled] Normalized ${events.length} events`);
  const newCount = await bulkUpsert(events);
  console.log(`[acled] Inserted ${newCount} new events`);

  if (newCount > 0) {
    broadcast({ type: "scraper_update", source: "acled", new_events: newCount });
  }

  return newCount;
}

function normalizeAcled(raw: any): EventCreate | null {
  const lat = parseFloat(raw.latitude);
  const lon = parseFloat(raw.longitude);
  if (!lat || !lon || (lat === 0 && lon === 0)) return null;

  const rawType = raw.event_type || "";
  const eventType = EVENT_TYPE_MAP[rawType] || rawType.toLowerCase().replace(/ /g, "_");

  const actors: string[] = [];
  if (raw.actor1) actors.push(raw.actor1);
  if (raw.actor2) actors.push(raw.actor2);

  const eventDate = new Date(raw.event_date);
  if (isNaN(eventDate.getTime())) return null;

  const locationParts: string[] = [];
  if (raw.location) locationParts.push(raw.location);
  if (raw.admin1) locationParts.push(raw.admin1);

  return {
    source: "acled",
    source_id: String(raw.data_id),
    event_type: eventType,
    sub_event_type: raw.sub_event_type || undefined,
    title: `${raw.sub_event_type || rawType} in ${raw.location || "Unknown"}`.slice(0, 512),
    description: raw.notes || undefined,
    url: raw.source_url || undefined,
    latitude: lat,
    longitude: lon,
    location_name: locationParts.join(", ") || undefined,
    country: raw.country || undefined,
    region: raw.region || undefined,
    event_date: eventDate,
    fatalities: parseInt(raw.fatalities) || 0,
    actors: actors.length > 0 ? actors : undefined,
    tags: rawType ? [rawType] : undefined,
  };
}
